### PR TITLE
Extras: Shell task

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
This PR makes it possible for user to write Shell tasks using python and interact with Flyte types.

Example of using the ShellTask
```python
t2 = ShellTask(
    name="task_2",
    debug=True,
    script="""
    cp {{ .inputs.x }} {{ .inputs.y }}
    tar -zcvf {{ .outputs.y }} {{ .inputs.y }}
    """,
    inputs=kwtypes(x=FlyteFile, y=FlyteDirectory),
    output_locs=[
        OutputLocation(var="y", var_type=FlyteFile, location="{{ .inputs.y }}.tar.gz")
    ],
)
```

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1788

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/extras/tasks/shell.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    _INPUT_REGEX = re.compile(r"({{\s*.inputs.(\w+)\s*}})", re.IGNORECASE)
+    _OUTPUT_REGEX = re.compile(r"({{\s*.outputs.(\w+)\s*}})", re.IGNORECASE)
+
+    def __init__(
+            self,
+            name: str,
+            debug: bool = False,
+            script: typing.Optional[str] = None,
+            script_file: typing.Optional[os.PathLike] = None,
+            task_config: T = None,
```
Comment (wild-endeavor) [inline flytekit/extras/tasks/shell.py:94]: I think intuitively this still doesn't really make sense to me.  Like I can put in a Python/Spark config here, but I can't put in a Hive task config right?  A Hive task config here just doesn't make sense.  The fact that some config types "make sense" and others don't make me feel like there's a missing interface here that we've missed.

File: flytekit/extras/tasks/shell.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    ):
+        """
+        Args:
+            name: str Name of the Task. Should be unique in the project
+            debug: bool Print the generated script and other debugging information
+            script: The actual script specified as a string
+            script_file: A path to the file that contains the script (Only script or script_file) can be provided
+            task_config: T Configuration for the task, can be either a Pod (or coming soon, BatchJob) config
+            inputs: A Dictionary of input names to types
+            output_locs: A list of :ref:class:`OutputLocations`
```
Comment (wild-endeavor) [inline flytekit/extras/tasks/shell.py:107]:
```suggestion
            output_locs: A list of :py:class:`OutputLocations`
```
?

File: flytekit/extras/tasks/shell.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                          f" Current directory contents: .\n-{fstr}")
+            raise
+
+        final_outputs = []
+        for v in self._output_locs:
+            if issubclass(v.var_type, FlyteFile):
+                final_outputs.append(FlyteFile(outputs[v.var]))
+            if issubclass(v.var_type, FlyteDirectory):
+                final_outputs.append(FlyteDirectory(outputs[v.var]))
+        return tuple(final_outputs)
```
Comment (wild-endeavor) [inline flytekit/extras/tasks/shell.py:205]: if there's just one output, do we still want to tuple?

**Review by wild-endeavor — COMMENTED** [review]
+1 when ready.

File: flytekit/extras/tasks/shell.py (hunk 3 above)
Comment (kumare3) [reply flytekit/extras/tasks/shell.py:205]: hmm good point

File: flytekit/extras/tasks/shell.py (hunk 2 above)
Comment (kumare3) [reply flytekit/extras/tasks/shell.py:107]: ya bad me

File: flytekit/extras/tasks/shell.py (hunk 1 above)
Comment (kumare3) [reply flytekit/extras/tasks/shell.py]: i think i will put a check for the supported config types using reflection?

Comment (kumare3) [reply flytekit/extras/tasks/shell.py]: done

File: flytekit/extras/tasks/shell.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        output_locs: typing.Optional[typing.List[OutputLocation]] = None,
+        **kwargs,
+    ):
+        """
+        Args:
+            name: str Name of the Task. Should be unique in the project
+            debug: bool Print the generated script and other debugging information
+            script: The actual script specified as a string
+            script_file: A path to the file that contains the script (Only script or script_file) can be provided
+            task_config: T Configuration for the task, can be either a Pod (or coming soon, BatchJob) config
```
Comment (eapolinario) [inline flytekit/extras/tasks/shell.py:105]: should this be restricted to pod and batch job?

File: flytekit/extras/tasks/shell.py (hunk 2 above)
Comment (eapolinario) [reply flytekit/extras/tasks/shell.py:107]: nit: still need to apply this change.

File: tests/flytekit/unit/extras/tasks/test_shell.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        output_locs=[
+            OutputLocation(var="x", var_type=FlyteDirectory, location="{{ .inputs.y }}"),
+            OutputLocation(var="y", var_type=FlyteFile, location="{{ .inputs.f }}.pyc"),
+        ],
+    )
+
+    assert t.script == s
+    x, y = t(f=test_csv, y=testdata, j=datetime.datetime(2021, 11, 10, 12, 15, 0))
+    assert x is not None
+    assert y.path[-4:] == ".pyc"
```
Comment (eapolinario) [inline tests/flytekit/unit/extras/tasks/test_shell.py:90]: could you also validate the contents of `y`?

**Review by eapolinario — COMMENTED** [review]
overall, LGTM, just a couple of minor comments / nits.

File: flytekit/extras/tasks/shell.py (hunk 4 above)
Comment (kumare3) [reply flytekit/extras/tasks/shell.py:105]: ya batch job config does not yet exist

File: tests/flytekit/unit/extras/tasks/test_shell.py (hunk 5 above)
Comment (kumare3) [reply tests/flytekit/unit/extras/tasks/test_shell.py:90]: sure

File: flytekit/extras/tasks/shell.py (hunk 4 above)
Comment (kumare3) [reply flytekit/extras/tasks/shell.py:105]: i did add the restriction, it was on my local good catch

File: tests/flytekit/unit/extras/tasks/test_shell.py (hunk 5 above)
Comment (kumare3) [reply tests/flytekit/unit/extras/tasks/test_shell.py:90]: cannot really, unless i guess i make a tempdir. maybe i should make a better test with tempdir

Comment (kumare3) [reply tests/flytekit/unit/extras/tasks/test_shell.py]: done

**Review by eapolinario — APPROVED** [review]

**Pull request merged**